# 2019-01-27

## 夜

<iframe height='405' width='590' frameborder='0' allowtransparency='true' scrolling='no' src='https://www.strava.com/activities/2106830698/embed/aeb9dc85666f078090e79c98c4d3290ebc13497b'></iframe>


## DRF Logging

- [python - How to make action logging in Django with Django Rest Framework - Stack Overflow](https://stackoverflow.com/questions/41556737/how-to-make-action-logging-in-django-with-django-rest-framework)
- [python - Logging requests to django-rest-framework - Stack Overflow](https://stackoverflow.com/questions/15578946/logging-requests-to-django-rest-framework/27928365)

HTTP_400_BAD_REQUEST:

- 通常、シリアライザ例外でリクエスト例外ではない ->  DRFの例外ハンドラが呼ばれない

### カスタムエラーハンドリング

app/exceptions.py:

~~~py
from rest_framework.views import exception_handler
from rest_framework.exceptions import ValidationError
from logging import getLogger
import json
import traceback

logger = getLogger('app')


def drf_exception_handler(exc, context):
    ''''
    # https://www.django-rest-framework.org/api-guide/exceptions/#custom-exception-handling
    context: 
        - view : DRF View Class
        - request: WSGIRequest
    '''
    for stack in traceback.format_stack():
        logger.debug(stack.strip())
    logger.debug(context)

    if isinstance(exc, ValidationError):
        # https://www.django-rest-framework.org/api-guide/exceptions/#validationerror
        logger.debug("* ValidationError:")
        logger.debug(json.dumps(exc.detail, ensure_ascii=False, indent=2))

    # response
    response = exception_handler(exc, context)
    return response
~~~

app/settings.py:

~~~py
REST_FRAMEWORK = {
    'EXCEPTION_HANDLER': 'app.exceptions.drf_exception_handler',
}
~~~

## 朝

<iframe height='405' width='590' frameborder='0' allowtransparency='true' scrolling='no' src='https://www.strava.com/activities/2105621837/embed/459cf4a3886de848792f9edb95b97bcbd3e6bf4e'></iframe>